### PR TITLE
[Scripts] Particle Explorer

### PR DESCRIPTION
**Pull request by imgntn**
This PR introduces a way to quickly explore the possibility space of our particle systems by linking dynamically created GUI with a particle system.  Run the script to spawn a particle system in front of you and use the GUI to make adjustments.  Export your settings as JSON to use in Interface, and import settings to load your system.

https://rawgit.com/imgntn/hifi/particle_explorer/examples/particle_explorer/particleExplorer.js

### CODE REVIEW

**Comment by imgntn** [comment]
@ericrius1 @huffman @ctrlaltdavid would you guys mind checking this out??  there's some improvements that can be made, aesthetically, if we like this approach.  also it can be set to send continuous updates instead of waiting until the final change is entered on a property, but i didn't want to overwhelm the entity server.  that way you wouldn't have to press enter to see updates.  we would want to debounce(throttle) it to some extent, i think?

**Comment by huffman** [comment]
I think we can throttle to a reasonable freq and still get good real-time results. edit.js sets a property and then immediately reads it again and updates the GUI (useful for updating the UI in case some constraint on the property was hit), so that would be something else to consider if you are trying to update a tool while you are still actively using it.

**Comment by imgntn** [comment]
note to self for monday: it could use some load testing on the 2-way part, since listening might be expensive...

File: examples/particle_explorer/generic.js [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+var controllers = [];
+var folders = [];
+var gui;
+var settings = new Settings();
+
+
+function convertBinaryToBoolean(value) {
+    if (value === 0) {
+        return false
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: all statements must end with ;

File: examples/particle_explorer/generic.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    } else {
+        return true
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: all statements must end with ;

File: examples/particle_explorer/generic.js [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+    //get object keys
+    var keys = _.keys(settings);
+
+    //for each key...
+    _.each(keys, function(key) {
+
+        // //FOR NOW TO SHOW THAT IT WORKS RESTRICT TO A FEW PROPERTIES.  NEED TO ADD SUPPORT FOR VEC3, QUAT, COLORS
+        if (key.indexOf('name') !== -1 || key.indexOf('description') !== -1 || key.indexOf('visible') !== -1 || key.indexOf('collisionsWillMove') !== -1) {
+
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: why so many blank lines?

File: examples/particle_explorer/generic.js [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+            // we aren't getting checkboxes for collisionsWillMove because it comes across the wire as a 1, not a true
+
+            if (key.indexOf('visible') !== -1) {
+                settings.visible = convertBinaryToBoolean(settings.visible)
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: all statements must end with ;

File: examples/particle_explorer/generic.js [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+            }
+            if (key.indexOf('collisionsWillMove') !== -1) {
+                settings.collisionsWillMove = convertBinaryToBoolean(settings.collisionsWillMove)
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: all statements must end with ;

File: examples/particle_explorer/generic.js [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+            }
+        } else {
+            return
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: all statements must end with ;

File: examples/particle_explorer/generic.js [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    gui.remember(settings);
+
+    //get object keys
+    var keys = _.keys(settings);
+
+    //for each key...
+    _.each(keys, function(key) {
+
+        // //FOR NOW TO SHOW THAT IT WORKS RESTRICT TO A FEW PROPERTIES.  NEED TO ADD SUPPORT FOR VEC3, QUAT, COLORS
+        if (key.indexOf('name') !== -1 || key.indexOf('description') !== -1 || key.indexOf('visible') !== -1 || key.indexOf('collisionsWillMove') !== -1) {
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: 3.4.1. The nominal case should be put in the if-part and the exception in the else-part of an if statement... Makes sure that the exceptions don't obscure the normal path of execution. This is important for both the readability and performance.

This if statement would make more sense if you tested for the "more likely case" of one of all of these values being not present, then you can simply "return" from the function... the code that's in your if clause, really belongs with the rest of the code in the function following the else case.

File: examples/particle_explorer/generic.js [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+
+}
+
+function writeVec3ToInterface(obj) {
+    print('VEC3 UPDATE TO INTERFACE FROM SETTINGS')
+    var sendData = {
+        messageType: "settings_update",
+        updatedSettings: obj,
+    }
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+function writeDataToInterface(property, value) {
+    console.log('WRITE SOME DATA TO INTERFACE' + property + ":::" + value)
+    var data = {};
+    data[property] = value;
+    var sendData = {
+        messageType: "settings_update",
+        updatedSettings: data,
+    }
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+    );
+
+
+}
+
+window.onload = function() {
+    console.log('GUI PAGE LOADED');
+
+    if (typeof EventBridge !== 'undefined') {
+        console.log('WE HAVE AN EVENT BRIDGE, SEND PAGE LOADED EVENT ')
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+        var stringifiedData = JSON.stringify({
+            messageType: 'page_loaded'
+        })
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+        console.log('SEND PAGE LOADED EVENT FROM GUI TO INTERFACE ')
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+        EventBridge.emitWebEvent(
+            stringifiedData
+        );
+        listenForSettingsUpdates();
+    } else {
+        console.log('No event bridge, probably not in interface.')
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+        data = JSON.parse(data);
+
+        if (data.messageType === 'object_update') {
+
+            _.each(data.objectSettings, function(value, key) {
+
+                settings[key] = value;
+
+                if (key.indexOf('visible') !== -1) {
+                    settings.visible = convertBinaryToBoolean(settings.visible)
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+
+                }
+                if (key.indexOf('collisionsWillMove') !== -1) {
+                    settings.collisionsWillMove = convertBinaryToBoolean(settings.collisionsWillMove)
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: blank line?

File: examples/particle_explorer/generic.js [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+                }
+
+            })
+
+        }
+        if (data.messageType === 'initial_settings') {
+            console.log('INITIAL SETTINGS FROM INTERFACE:::' + JSON.stringify(data.initialSettings))
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+            _.each(data.initialSettings, function(value, key) {
+                console.log('key  ' + key);
+                console.log('value  ' + JSON.stringify(value));
+                settings[key] = {};
+                settings[key] = value;
+            })
+            loadGUI();
+        }
+        if (data.messageType === 'settings_update') {
+            console.log('SETTINGS UPDATE FROM INTERFACE:::' + JSON.stringify(data.updatedSettings))
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/generic.js [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+            _.each(data.updatedSettings, function(value, key) {
+                console.log('setting,value', setting, value)
```
Comment (ZappoMan) [inline examples/particle_explorer/generic.js]: coding standard: statements should end with a ;

File: examples/particle_explorer/genericProperties.js [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+
+var box = Entities.addEntity({
+    type: 'Box',
+    visible: true,
+    collisionsWillMove: true,
+    color: {
+        red: 0,
+        green: 255,
+        blue: 0
+
```
Comment (ZappoMan) [inline examples/particle_explorer/genericProperties.js]: extra line?

File: examples/particle_explorer/genericProperties.js [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+SettingsWindow = function() {
+    var _this = this;
+
+    this.webWindow = null;
+
+    this.init = function() {
+        _this.webWindow = new WebWindow('genericProperties', Script.resolvePath('index.html'), 400, 600, true);
+        _this.webWindow.setVisible(true);
+        _this.webWindow.eventBridge.webEventReceived.connect(_this.onWebEventReceived);
+        Script.update.connect(waitForObjectAuthorization)
```
Comment (ZappoMan) [inline examples/particle_explorer/genericProperties.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/genericProperties.js [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+    };
+
+    function waitForObjectAuthorization() {
+        var properties = Entities.getEntityProperties(box, "isKnownID");
+        var isKnownID = properties.isKnownID
```
Comment (ZappoMan) [inline examples/particle_explorer/genericProperties.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/genericProperties.js [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+}
+
+function sendUpdatedObject(properties) {
+    // print('SENDING UPDATED OBJECT FROM INTERFACE');
+    var settings = {
+        messageType: 'object_update',
+        objectSettings: properties
+    };
+    settingsWindow.sendData(settings)
```
Comment (ZappoMan) [inline examples/particle_explorer/genericProperties.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/genericProperties.js [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+function editEntity(properties) {
+    Entities.editEntity(box, properties);
+    var currentProperties = Entities.getEntityProperties(box);
+    settingsWindow.sendData({
+        messageType: 'settings_update',
+        updatedSettings: currentProperties
+    })
```
Comment (ZappoMan) [inline examples/particle_explorer/genericProperties.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+//  todo: folders, color pickers, animation settings, scale gui width with window resizing  
+//
+
+var PI = 3.141593,
+    DEG_TO_RAD = PI / 180.0;
+
+function radiansToDegrees(radians) {
+    return radians * (180 / PI)
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 26]
```diff
@@ -0,0 +1,217 @@
+//
+//  main.js
+//  
+//
+//  Created by James B. Pollack @imgntnon 9/26/2015
+//  Copyright 2014 High Fidelity, Inc.
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: 2014?

File: examples/particle_explorer/main.js [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+var groups = [
+    'accelerationSpread',
+    'color',
+    'colorSpread',
+    'colorStart',
+    'colorFinish',
+    'emitAcceleration',
+    'emitDimensions',
+    'emitOrientation'
+]
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+    //if not autoplacing, put gui in a custom container
+    if (gui.autoPlace === false) {
+
+        var customContainer = document.getElementById('my-gui-container');
+        customContainer.appendChild(gui.domElement);
+        gui.width = 400;
+
+    }
+
+var folder = gui.addFolder('thingo')
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+folder.add(particleExplorer.dimensions,'x')
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 30]
```diff
@@ -0,0 +1,217 @@
...
+folder.add(particleExplorer.dimensions,'y')
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+folder.add(particleExplorer.dimensions,'z')
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 32]
```diff
@@ -0,0 +1,217 @@
...
+        //hook into change events for this gui controller
+        controller.onChange(function(value) {
+            // Fires on every change, drag, keypress, etc.
+            // writeDataToInterface(this.property, value, this.shouldGroup)
+        });
+
+        controller.onFinishChange(function(value) {
+            // console.log('should group?', controller.shouldGroup)
+            // Fires when a controller loses focus.
+            writeDataToInterface(this.property, value, this.shouldGroup)
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 33]
```diff
@@ -0,0 +1,217 @@
...
+    var groupProperty = null;
+    var groupProperties = null;
+
+    if (shouldGroup) {
+        var separated = property.split("_");
+        group = separated[0];
+        groupProperty = separated[1];
+        var groupProperties = {}
+        groupProperties[group] = {};
+        groupProperties[group][groupProperty] = value
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 34]
```diff
@@ -0,0 +1,217 @@
...
+    // console.log('property, value, shouldGroup',property, value, shouldGroup)
+    var group = null;
+    var groupProperty = null;
+    var groupProperties = null;
+
+    if (shouldGroup) {
+        var separated = property.split("_");
+        group = separated[0];
+        groupProperty = separated[1];
+        var groupProperties = {}
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 35]
```diff
@@ -0,0 +1,217 @@
...
+        groupProperties[group][groupProperty] = value
+            // console.log(groupProperties)
+    }
+
+    var data = {
+        type: "particleExplorer_update",
+        shouldGroup: shouldGroup,
+        groupProperties: groupProperties,
+        singleProperty: {}
+    }
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 36]
```diff
@@ -0,0 +1,217 @@
...
+
+    data['singleProperty'][property] = value
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 37]
```diff
@@ -0,0 +1,217 @@
...
+
+    var stringifiedData = JSON.stringify(data)
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/main.js [hunk 38]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+}
+
+function listenForSettingsUpdates() {
+    if (typeof EventBridge !== 'undefined') {
+        EventBridge.scriptEventReceived.connect(function(data) {
+            data = JSON.parse(data);
+            if (data.type === 'particleSettingsUpdate') {
+                var particleSettings = data.particleSettings
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 39]
```diff
@@ -0,0 +1,189 @@
...
+var testParticles = Entities.addEntity(startingParticles);
+
+SettingsWindow = function() {
+    var _this = this;
+    this.webWindow = null;
+    this.init = function() {
+        _this.webWindow = new WebWindow('ParticleExplorer', Script.resolvePath('index.html'), 400, 600, true);
+        _this.webWindow.setVisible(true);
+        _this.webWindow.eventBridge.webEventReceived.connect(_this.onWebEventReceived);
+        print('INIT testParticles' + testParticles)
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 40]
```diff
@@ -0,0 +1,189 @@
...
+    };
+    this.sendData = function(data) {
+        _this.webWindow.eventBridge.emitScriptEvent(JSON.stringify(data));
+    };
+    this.onWebEventReceived = function(data) {
+        // print('DATA ' + data)
+        var _data = JSON.parse(data)
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 41]
```diff
@@ -0,0 +1,189 @@
...
+        if (_data.type !== 'particleExplorer_update') {
+            return;
+        }
+        if (_data.shouldGroup === true) {
+            // print('USE GROUP PROPERTIES')
+            editEntity(_data.groupProperties)
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 42]
```diff
@@ -0,0 +1,189 @@
...
+            return;
+        } else {
+            // print('USE A SINGLE PROPERTY')
+            editEntity(_data.singleProperty)
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 43]
```diff
@@ -0,0 +1,189 @@
...
+
+
+    };
+
+}
+
+
+function editEntity(properties) {
+    Entities.editEntity(testParticles, properties);
+    var currentProperties = Entities.getEntityProperties(testParticles)
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

File: examples/particle_explorer/particleExplorer.js [hunk 44]
```diff
@@ -0,0 +1,189 @@
...
+    // print('CURRENT PROPS', JSON.stringify(currentProperties))
+    settingsWindow.sendData({type:'particleSettingsUpdate',particleSettings:currentProperties})
```
Comment (ZappoMan) [inline examples/particle_explorer/particleExplorer.js]: coding standard: all statements should end with a ;

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7c9d7ffefd4c8f065ac7d6a81dbeca56fe917b44 available at: https://gist.github.com/238a6fc834411eba1870<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ea20fc97867ea14da747b9272b3351d745bc2da1 available at: https://gist.github.com/238a6fc834411eba1870<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d1cbcf5554c73c103fe0ab789c0d7732e3c2a081 available at: https://gist.github.com/238a6fc834411eba1870<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5909.dmg

**Comment by ZappoMan** [comment]
@imgntn - something is wrong with this PR, it shows C++ code being changed, I think you need to merge with upstream master or rebase. Can you please do that?

File: examples/particle_explorer/main.js [hunk 45]
```diff
@@ -0,0 +1,526 @@
...
+    folders.push(folder);
+    folder.open();
+}
+
+function convertColorObjectToArray(colorObject) {
+    var colorArray = [];
+    _.each(colorObject, function(singleColor) {
+        colorArray.push(singleColor);
+    })
+    return colorArray
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: statements end with ;

File: examples/particle_explorer/main.js [hunk 46]
```diff
@@ -0,0 +1,526 @@
...
+}
+
+function convertColorArrayToObject(colorArray) {
+    var colorObject = {
+        red: colorArray[0],
+        green: colorArray[1],
+        blue: colorArray[2]
+    }
+    return colorObject
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: statements end with ;

File: examples/particle_explorer/main.js [hunk 47]
```diff
@@ -0,0 +1,526 @@
...
+    })
+    return colorArray
+}
+
+function convertColorArrayToObject(colorArray) {
+    var colorObject = {
+        red: colorArray[0],
+        green: colorArray[1],
+        blue: colorArray[2]
+    }
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: statements end with ;

File: examples/particle_explorer/main.js [hunk 48]
```diff
@@ -0,0 +1,526 @@
...
+    });
+    folders.push(folder);
+    folder.open();
+}
+
+function convertColorObjectToArray(colorObject) {
+    var colorArray = [];
+    _.each(colorObject, function(singleColor) {
+        colorArray.push(singleColor);
+    })
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: statements end with ;

**Comment by imgntn** [comment]
i merged the upstream/master and pushed again.  i'm not sure why it shows all the code being changed -- i haven't been on the C side of things at all.  that must be the result of previous merges.  not sure why it shows up like that @ZappoMan

File: examples/particle_explorer/main.js [hunk 49]
```diff
@@ -0,0 +1,526 @@
...
+    var data = {};
+    data[property] = value;
+    var sendData = {
+        messageType: "settings_update",
+        updatedSettings: data,
+    };
+
+    var stringifiedData = JSON.stringify(sendData);
+
+    EventBridge.emitWebEvent(
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: really??? why 3 lines... please make this 1 line.

File: examples/particle_explorer/main.js [hunk 50]
```diff
@@ -0,0 +1,526 @@
...
+
+function writeVec3ToInterface(obj) {
+    var sendData = {
+        messageType: "settings_update",
+        updatedSettings: obj,
+    };
+
+    var stringifiedData = JSON.stringify(sendData);
+
+    EventBridge.emitWebEvent(
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: one line please

File: examples/particle_explorer/main.js [hunk 51]
```diff
@@ -0,0 +1,526 @@
...
+    } catch (e) {
+        copyElement.remove();
+        console.log("document.execCommand('copy'); is not supported");
+        prompt('Copy the text below. (ctrl c, enter)', text);
+    } finally {
+        if (typeof e == 'undefined') {
+            copyElement.remove();
+        }
+    }
+    console.log('should have clipped text')
```
Comment (ZappoMan) [inline examples/particle_explorer/main.js]: coding standard: statements end with ;

File: examples/particle_explorer/main.js (hunk 49 above)
Comment (imgntn) [inline examples/particle_explorer/main.js]: why?  is it not coding-standard compliant?  the trailing comma should be removed from updatedSettings but looks good to me other than that

Comment (ZappoMan) [inline examples/particle_explorer/main.js]: it's wasteful - makes the file longer and harder to read. there's nothing explicit in the coding standard about this...

**Pull request merged**